File: mobile/android/app/src/main/java/com/nextgenz/keyboard/PredictionEngine.kt
package com.nextgenz.keyboard

import android.content.Context

class PredictionEngine(context: Context) {
    private val prefixMap = HashMap<String, List<String>>()
    private val nextMap = HashMap<String, List<String>>()

    init {
        loadPrefix(context)
        loadNext(context)
    }

    fun suggestPrefix(prefix: String, topN: Int = 5): List<String> {
        if (prefix.isBlank() || !isKhmerWord(prefix)) return emptyList()
        return prefixMap[prefix]?.take(topN) ?: emptyList()
    }

    fun predictNext(w1: String, w2: String, topN: Int = 5): List<String> {
        if (w1.isBlank() || w2.isBlank()) return emptyList()
        if (!isKhmerWord(w1) || !isKhmerWord(w2)) return emptyList()
        val key = "$w1\t$w2"
        return nextMap[key]?.take(topN) ?: emptyList()
    }

    private fun loadPrefix(context: Context) {
        context.assets.open("model/prefix.tsv").bufferedReader(Charsets.UTF_8).useLines { lines ->
            lines.forEach { line ->
                val firstTab = line.indexOf('\t')
                if (firstTab <= 0) return@forEach
                val key = line.substring(0, firstTab)
                val values = line.substring(firstTab + 1).split('|').filter { it.isNotBlank() }
                if (isKhmerWord(key) && values.isNotEmpty()) {
                    val khmerValues = values.filter { isKhmerWord(it) }
                    if (khmerValues.isEmpty()) return@forEach
                    prefixMap[key] = khmerValues
                }
            }
        }
    }

    private fun loadNext(context: Context) {
        context.assets.open("model/next.tsv").bufferedReader(Charsets.UTF_8).useLines { lines ->
            lines.forEach { line ->
                val parts = line.split('\t')
                if (parts.size < 3) return@forEach
                val key = "${parts[0]}\t${parts[1]}"
                val values = parts[2].split('|').filter { it.isNotBlank() }
                if (isKhmerWord(parts[0]) && isKhmerWord(parts[1]) && values.isNotEmpty()) {
                    val khmerValues = values.filter { isKhmerWord(it) }
                    if (khmerValues.isEmpty()) return@forEach
                    nextMap[key] = khmerValues
                }
            }
        }
    }

    private fun isKhmerWord(text: String): Boolean {
        if (text.isBlank()) return false
        var hasKhmer = false
        for (ch in text) {
            if (ch.isWhitespace() || ch == '\'' || ch == '’' || ch == '-' || ch == '_') {
                continue
            }
            val isKhmer = (ch.code in 0x1780..0x17FF) || (ch.code in 0x19E0..0x19FF)
            if (!isKhmer) return false
            hasKhmer = true
        }
        return hasKhmer
    }
}

File: mobile/ios/Shared/PredictionEngine.swift
import Foundation

final class PredictionEngine {
    private var prefixMap: [String: [String]] = [:]
    private var nextMap: [String: [String]] = [:]

    func loadModel(bundle: Bundle = .main) {
        prefixMap = loadPrefix(bundle: bundle)
        nextMap = loadNext(bundle: bundle)
    }

    func suggestPrefix(_ prefix: String, topN: Int = 3) -> [String] {
        guard Self.isKhmerWord(prefix) else { return [] }
        return Array((prefixMap[prefix] ?? []).prefix(max(1, topN)))
    }

    func predictNext(_ w1: String, _ w2: String, topN: Int = 3) -> [String] {
        guard Self.isKhmerWord(w1), Self.isKhmerWord(w2) else { return [] }
        return Array((nextMap["\(w1)\t\(w2)"] ?? []).prefix(max(1, topN)))
    }

    static func isKhmerWord(_ text: String) -> Bool {
        guard !text.isEmpty else { return false }
        var hasKhmer = false
        for scalar in text.unicodeScalars {
            let v = scalar.value
            if CharacterSet.whitespacesAndNewlines.contains(scalar) { continue }
            let isKhmer = (0x1780...0x17FF).contains(v) || (0x19E0...0x19FF).contains(v)
            if !isKhmer { return false }
            hasKhmer = true
        }
        return hasKhmer
    }

    private func loadPrefix(bundle: Bundle) -> [String: [String]] {
        guard let url = bundle.url(forResource: "prefix", withExtension: "tsv", subdirectory: "Model"),
              let text = try? String(contentsOf: url, encoding: .utf8) else {
            return [:]
        }
        var out: [String: [String]] = [:]
        for line in text.split(separator: "\n") {
            let parts = line.split(separator: "\t", maxSplits: 1)
            if parts.count != 2 { continue }
            let key = String(parts[0])
            if !Self.isKhmerWord(key) { continue }
            let vals = parts[1].split(separator: "|").map(String.init).filter { Self.isKhmerWord($0) }
            if !vals.isEmpty { out[key] = vals }
        }
        return out
    }

    private func loadNext(bundle: Bundle) -> [String: [String]] {
        guard let url = bundle.url(forResource: "next", withExtension: "tsv", subdirectory: "Model"),
              let text = try? String(contentsOf: url, encoding: .utf8) else {
            return [:]
        }
        var out: [String: [String]] = [:]
        for line in text.split(separator: "\n") {
            let parts = line.split(separator: "\t")
            if parts.count < 3 { continue }
            let w1 = String(parts[0])
            let w2 = String(parts[1])
            if !Self.isKhmerWord(w1) || !Self.isKhmerWord(w2) { continue }
            let vals = parts[2].split(separator: "|").map(String.init).filter { Self.isKhmerWord($0) }
            if !vals.isEmpty { out["\(w1)\t\(w2)"] = vals }
        }
        return out
    }
}

